# Week 4: Vector Representations & Similarity - Homework

**ML2: Advanced Machine Learning**

**Estimated Time**: 1 hour

---

This homework combines programming exercises and knowledge-based questions to reinforce this week's concepts.

## Setup

Run this cell to import necessary libraries:

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import torch
import torch.nn as nn

# Set random seed for reproducibility
np.random.seed(42)
torch.manual_seed(42)

print('✓ Libraries imported successfully')

✓ Libraries imported successfully


---
## Part 1: Programming Exercises (60%)

Complete the following programming tasks. Read each description carefully and implement the requested functionality.

### Exercise 1: Experiment: Cosine vs Euclidean Similarity

**Time**: 10 min

Observe how cosine similarity and Euclidean distance behave differently, especially with vector magnitude.

In [2]:
import numpy as np

# User preferences (ratings 1-5 for 5 movies)
user_a = np.array([5, 5, 1, 1, 1])  # Loves action, hates romance
user_b = np.array([4, 4, 1, 1, 1])  # Same pattern, slightly lower ratings
user_c = np.array([1, 1, 5, 5, 5])  # Opposite preferences

def cosine_similarity(v1, v2):
    return np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2))

def euclidean_distance(v1, v2):
    return np.linalg.norm(v1 - v2)

print("User A vs User B:")
print(f"  Cosine similarity: {cosine_similarity(user_a, user_b):.4f}")
print(f"  Euclidean distance: {euclidean_distance(user_a, user_b):.4f}")

print("\nUser A vs User C:")
print(f"  Cosine similarity: {cosine_similarity(user_a, user_c):.4f}")
print(f"  Euclidean distance: {euclidean_distance(user_a, user_c):.4f}")

# Now scale user_b by 2x (enthusiastic rater)
user_b_scaled = user_b * 2

print("\nUser A vs User B (2x scaled):")
print(f"  Cosine similarity: {cosine_similarity(user_a, user_b_scaled):.4f}")
print(f"  Euclidean distance: {euclidean_distance(user_a, user_b_scaled):.4f}")

# TODO: Answer reflection questions about what this tells you

User A vs User B:
  Cosine similarity: 0.9984
  Euclidean distance: 1.4142

User A vs User C:
  Cosine similarity: 0.3913
  Euclidean distance: 8.9443

User A vs User B (2x scaled):
  Cosine similarity: 0.9984
  Euclidean distance: 4.5826


### Exercise 2: Experiment: Sparsity Problem

**Time**: 8 min

See how sparse vectors (mostly zeros) make similarity search difficult.

In [3]:
import numpy as np

np.random.seed(42)

# Sparse binary vectors (1000 dimensions, only 10 are 1)
def create_sparse_vector(dim=1000, num_ones=10):
    vec = np.zeros(dim)
    indices = np.random.choice(dim, num_ones, replace=False)
    vec[indices] = 1
    return vec

# Create 5 users with sparse preferences
users = [create_sparse_vector() for _ in range(5)]

# Compute pairwise similarities
print("Pairwise Cosine Similarities:")
for i in range(len(users)):
    for j in range(i+1, len(users)):
        sim = np.dot(users[i], users[j]) / (np.linalg.norm(users[i]) * np.linalg.norm(users[j]))
        overlap = int(np.dot(users[i], users[j]))  # Number of shared 1s
        print(f"User {i} vs User {j}: similarity={sim:.4f}, shared items={overlap}")

# TODO: What pattern do you notice? Why is this a problem?

Pairwise Cosine Similarities:
User 0 vs User 1: similarity=0.0000, shared items=0
User 0 vs User 2: similarity=0.0000, shared items=0
User 0 vs User 3: similarity=0.0000, shared items=0
User 0 vs User 4: similarity=0.0000, shared items=0
User 1 vs User 2: similarity=0.0000, shared items=0
User 1 vs User 3: similarity=0.0000, shared items=0
User 1 vs User 4: similarity=0.0000, shared items=0
User 2 vs User 3: similarity=0.0000, shared items=0
User 2 vs User 4: similarity=0.0000, shared items=0
User 3 vs User 4: similarity=0.0000, shared items=0


---
## Part 2: Knowledge Questions (40%)

Answer the following questions to test your conceptual understanding.

### Question 1 (Short Answer)

**Question 1 - Cosine vs Euclidean Conceptual Understanding**

Cosine similarity measures the ANGLE between vectors.
Euclidean distance measures the MAGNITUDE of difference.

Explain:
1. Why is cosine similarity better for comparing user preferences?
2. Give an example where two users have the same taste but different cosine vs Euclidean similarity.
3. What does a cosine similarity of 1.0 mean? What about -1.0?

**Hint**: Think about rating scales. One user might rate everything 1-5, another 2-4, but have the same preferences.

**Your Answer**:

Cosine similarity captures the pattern of preferences regardless of how extreme the ratings are, so a lenient rater and a harsh rater with identical tastes will still show high similarity. For example, User A rating [5,5,1] and User B rating [10,10,2] have cosine=1.0 but a large Euclidean distance. Cosine=1.0 means identical direction (same taste), -1.0 means exactly opposite.

### Question 2 (Short Answer)

**Question 2 - Experiment Reflection: Scaling**

In the 'Cosine vs Euclidean' experiment, User B was scaled by 2x (all ratings doubled).

Observe what happened to:
1. Cosine similarity (did it change?)
2. Euclidean distance (did it change?)
3. Which metric correctly recognizes that User A and User B (scaled) have the SAME preferences?

**Hint**: Cosine is scale-invariant. It only cares about direction (pattern), not magnitude (enthusiasm).

**Your Answer**:

Cosine similarity stayed at 0.9984 after scaling, because doubling all ratings doesn't change the direction of the vector. Euclidean distance jumped from 1.41 to 4.58, making B look less similar to A even though nothing changed about their preferences. Cosine correctly recognizes they have the same taste.

### Question 3 (Multiple Choice)

**Question 3 - When to Use Which Metric**

You're building a movie recommendation system. Users rate movies 1-5 stars. Some users are "easy graders" (average 4.0) and some are "harsh critics" (average 2.5), but they might have similar taste.

Which similarity metric should you use?

A) Euclidean distance
B) Cosine similarity
C) Manhattan distance
D) Doesn't matter

A) Euclidean distance
B) Cosine similarity
C) Manhattan distance
D) Doesn't matter

**Hint**: You want to capture PREFERENCE PATTERNS, not absolute rating levels.

**Your Answer**: B

**Explanation**: Cosine similarity ignores the overall rating level and focuses on which movies a user liked relative to others, so an easy grader and a harsh critic with the same taste will still match.

### Question 4 (Short Answer)

**Question 4 - The Sparsity Problem**

Based on the 'Sparsity Problem' experiment:

With 1000 possible items and users only rating 10 items each, most users had NO shared items (overlap=0).

Explain:
1. Why does sparsity make similarity computation difficult?
2. What happens to cosine similarity when two vectors share no non-zero elements?
3. How might you solve this problem in practice?

**Hint**: If vectors don't overlap, you can't find similarity even if users actually have similar tastes.

**Your Answer**:

Every pair in the experiment had similarity=0.0 because none shared any rated items, making the dot product zero. Two users with identical taste but no overlapping items are indistinguishable from two users with opposite taste. The solution is to learn dense embeddings (via matrix factorization or neural nets) that represent latent preferences rather than raw item ratings.

### Question 5 (Short Answer)

**Question 5 - Dimensionality Reduction Motivation**

High-dimensional sparse vectors → hard to find similarities
Low-dimensional dense vectors → easier to find patterns

Explain: How could you transform high-dimensional sparse movie ratings into low-dimensional dense embeddings? What would the embeddings capture?

**Hint**: Think about learning latent factors like 'action fan' or 'romance lover' instead of storing raw ratings.

**Your Answer**:

Matrix factorization or an autoencoder can compress sparse rating vectors into dense low-dimensional embeddings. The dimensions of the learned space represent latent factors like genre preferences or style affinity, which are shared across users even if their specific rated items don't overlap.

### Question 6 (Multiple Choice)

**Question 6 - Curse of Dimensionality**

In very high dimensions (e.g., 10,000), strange things happen to distances. Most points appear roughly equidistant from each other.

Why is this a problem for nearest neighbor search?

A) It makes computation slow
B) It makes all similarities look the same (less informative)
C) It uses too much memory
D) It requires more training data

A) It makes computation slow
B) It makes all similarities look the same (less informative)
C) It uses too much memory
D) It requires more training data

**Hint**: When everything is equally far apart, you can't distinguish near from far.

**Your Answer**: B

**Explanation**: In high dimensions, pairwise distances concentrate around the same value, so nearest neighbors are barely closer than random points and the ranking loses meaning.

### Question 7 (Short Answer)

**Question 7 - Learned Representations vs Manual Features**

Manual approach: Design features like "likes action", "likes romance" (requires domain expertise)
Learned approach: Let a neural network discover features automatically from data

Explain:
1. What advantage does the learned approach have?
2. What disadvantage might it have?
3. When would you prefer manual features?

**Hint**: Learned = automatic but less interpretable. Manual = interpretable but requires expertise.

**Your Answer**:

Learned features scale automatically and can discover patterns that domain experts wouldn't think to encode. The downside is they're harder to interpret and require more data to learn from. Manual features are preferable when interpretability is required, data is scarce, or the domain structure is well understood.

### Question 8 (Short Answer)

**Question 8 - Dot Product Interpretation**

Cosine similarity = (A · B) / (||A|| × ||B||)

The numerator is the dot product A · B.

Explain in intuitive terms: What does a high dot product between two vectors mean? What about a dot product of zero?

**Hint**: Dot product measures alignment. High = pointing same direction. Zero = perpendicular.

**Your Answer**:

A high dot product means the vectors are pointing in the same direction, i.e., the users have similar preferences across the same dimensions. A dot product of zero means the vectors are perpendicular and have no alignment at all, which in a sparse setting often just means no shared rated items rather than truly opposite tastes.

### Question 9 (Short Answer)

**Question 9 - Normalization Impact**

If you normalize all vectors to unit length (magnitude 1), what happens to the relationship between cosine similarity and Euclidean distance?

Hint: Try the math with ||A|| = ||B|| = 1

**Hint**: When vectors are normalized, cosine similarity and Euclidean distance become related.

**Your Answer**:

When ||A|| = ||B|| = 1, the Euclidean distance squared is ||A - B||² = 2 - 2(A·B) = 2 - 2*cosine. So Euclidean distance = sqrt(2(1 - cosine)), meaning they become equivalent measures and ranking by one is the same as ranking by the other.

### Question 10 (Short Answer)

**Question 10 - Real-World Application**

Spotify has 100 million songs and wants to find "similar songs" for recommendations.

Explain:
1. Why can't they store a 100M × 100M similarity matrix?
2. How do learned embeddings (e.g., 128-dimensional vectors per song) solve this?
3. What tradeoff are they making?

**Hint**: 100M × 100M floats = massive storage. 100M × 128 floats = much smaller. But you lose some information in compression.

**Your Answer**:

A 100M x 100M matrix of floats would be around 40 petabytes, which is infeasible to store or query. Storing 128-dimensional embeddings per song is roughly 50GB and similarity can be computed on the fly via dot products. The tradeoff is that embeddings are an approximation, so similar songs in raw feature space might not always be neighbors in the learned space.

---
## Submission

Before submitting:
1. Run all cells to ensure code executes without errors
2. Check that all questions are answered
3. Review your explanations for clarity

**To Submit**:
- File → Download → Download .ipynb
- Submit the notebook file to your course LMS

**Note**: Make sure your name is in the filename (e.g., homework_01_yourname.ipynb)